packages/arb-ts/src/lib/l1Bridge.ts

```diff
@@ -102,82 +109,60 @@ export class L1Bridge {
+    const contractCode = await this.l1Provider.getCode(contractAddress)
+    return !(contractCode.length > 2)
+  }
+
+  public async getL1TokenData(erc20L1Address: string): Promise<L1TokenData> {
     const walletAddress = await this.getWalletAddress()
     const gatewayAddress = await this.getGatewayAddress(erc20L1Address)
 
-    if (!tokenData.ERC20) {
-      if ((await this.l1Provider.getCode(erc20L1Address)).length > 2) {
-        // If this will throw if not an ERC20, which is what we *want*.
-        const ethERC20TokenContract = await ERC20__factory.connect(
-          erc20L1Address,
-          this.l1Signer
-        )
-        const [balance] = await ethERC20TokenContract.functions.balanceOf(
-          walletAddress
-        )
+    if (!this.contractExists(erc20L1Address))
+      throw new Error(`No ERC20 at ${erc20L1Address} `)
 
-        const [allowance] = await ethERC20TokenContract.functions.allowance(
-          walletAddress,
-          gatewayAddress
-        )
-        // non-standard
-        const symbol = await ethERC20TokenContract.functions
-          .symbol()
-          .then(([res]) => res)
-          .catch(() => addressToSymbol(erc20L1Address))
+    // If this will throw if not an ERC20, which is what we *want*.
+    const ethERC20TokenContract = await ERC20__factory.connect(
+      erc20L1Address,
+      this.l1Signer
+    )
+    const [balance] = await ethERC20TokenContract.functions.balanceOf(
+      walletAddress
+    )
 
-        const decimals = await ethERC20TokenContract.functions
-          .decimals()
-          .then(([res]) => res)
-          .catch(() => 18)
+    const [allowance] = await ethERC20TokenContract.functions.allowance(
+      walletAddress,
+      gatewayAddress
+    )
+    // non-standard
+    const symbol = await ethERC20TokenContract.functions
+      .symbol()
+      .then(([res]) => res)
+      .catch(() => addressToSymbol(erc20L1Address))
 
-        const name = await ethERC20TokenContract.functions
-          .name()
-          .then(([res]) => res)
-          .catch(() => symbol + '_Token')
-        const allowanceLimit = BigNumber.from(
-          '0xffffffffffffffffffffffff'
-        ) /** for ERC20s that cap approve at 96 bits  */
-        const allowed = await allowance.gte(allowanceLimit.div(2))
-        tokenData.ERC20 = {
-          contract: ethERC20TokenContract,
-          balance,
-          allowed,
-          symbol,
-          decimals,
-          name,
-        }
-      } else {
-        throw new Error(`No ERC20 at ${erc20L1Address} `)
-      }
-    } else {
-      const ethERC20TokenContract = await ERC20__factory.connect(
-        erc20L1Address,
-        this.l1Signer
-      )
-      const [balance] = await ethERC20TokenContract.functions.balanceOf(
-        walletAddress
-      )
-      tokenData.ERC20.balance = balance
+    const decimals = await ethERC20TokenContract.functions
+      .decimals()
+      .then(([res]) => res)
+      .catch(() => 18)
```

Comment:
Would be good to pipeline all of these calls, creating promises and then `await`ing

Reply:
how about we multicall them